pkg/controller/backup_storage_location_controller.go

```diff
@@ -0,0 +1,210 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"encoding/json"
+	"strings"
+	"time"
+
+	jsonpatch "github.com/evanphx/json-patch"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/types"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	velerov1api "github.com/vmware-tanzu/velero/pkg/apis/velero/v1"
+	velerov1client "github.com/vmware-tanzu/velero/pkg/generated/clientset/versioned/typed/velero/v1"
+	velerov1listers "github.com/vmware-tanzu/velero/pkg/generated/listers/velero/v1"
+	"github.com/vmware-tanzu/velero/pkg/persistence"
+	"github.com/vmware-tanzu/velero/pkg/plugin/clientmgmt"
+)
+
+type backupStorageLocationController struct {
+	*genericController
+
+	namespace                    string
+	defaultBackupLocation        string
+	defaultStoreValidationPeriod time.Duration
+	backupLocationClient         velerov1client.BackupStorageLocationsGetter
+	backupStorageLocationLister  velerov1listers.BackupStorageLocationLister
+	newPluginManager             func(logrus.FieldLogger) clientmgmt.Manager
+	newBackupStore               func(*velerov1api.BackupStorageLocation, persistence.ObjectStoreGetter, logrus.FieldLogger) (persistence.BackupStore, error)
+}
+
+func NewBackupStorageLocationController(
+	namespace string,
+	defaultBackupLocation string,
+	defaultStoreValidationPeriod time.Duration,
+	backupLocationClient velerov1client.BackupStorageLocationsGetter,
+	backupStorageLocationLister velerov1listers.BackupStorageLocationLister,
+	newPluginManager func(logrus.FieldLogger) clientmgmt.Manager,
+	logger logrus.FieldLogger,
+) Interface {
+	if defaultStoreValidationPeriod <= 0 {
+		defaultStoreValidationPeriod = time.Minute
+	}
+
+	logger.Infof("Backup location store validation period is %v", defaultStoreValidationPeriod)
+
+	c := backupStorageLocationController{
+		genericController:            newGenericController("backup-storage-location", logger),
+		namespace:                    namespace,
+		defaultBackupLocation:        defaultBackupLocation,
+		defaultStoreValidationPeriod: defaultStoreValidationPeriod,
+		backupLocationClient:         backupLocationClient,
+		backupStorageLocationLister:  backupStorageLocationLister,
+
+		// use variables to refer to these functions so they can be
+		// replaced with fakes for testing.
+		newPluginManager: newPluginManager,
+		newBackupStore:   persistence.NewObjectBackupStore,
+	}
+
+	c.resyncFunc = c.run
```

Comment:
I think in addition to the periodic validation via this resync func, we should also add OnAdd/OnUpdate event handlers to the backup storage location informer so that whenever a BSL is created or updated, it immediately triggers a validation. WDYT?

Reply:
Yes, good idea.

Would we need both the handlers and the BSL validation though? 

I'm thinking add the validation to the handlers (create/update), and keep the BSL looping and checking the CR field just so it does the proper periodic logging. Would this work?

Reply:
> This brings up another nuance, which is - if we add a validation check in the backup controller to ensure the BSL is "Available" - should we rely on the CR field, or should we actually try to connect to the object store (e.g. via the backupStore.IsValid() method)? Relying on the CR field is easier/more efficient, but it could theoretically be out-of-date.

If we have the handler validating right on `update` then it will never be out-of-date. Right?

Reply:
I think it makes sense to do the validation both when a BSL is added/updated, *and* periodically (via the resync func). That way, we immediately get an updated phase when the BSL is created or changed, but we are also regularly checking to ensure we can still connect to the object store.

Reply:
It could still be out-of-date: the minio instance goes down, the S3 credentials are expired, etc. None of those would involve a change to the BSL, but could still mean that it can't be connected to.

Reply:
Yes, good point!